lib/dbus/value.rb

```diff
@@ -0,0 +1,700 @@
+# frozen_string_literal: true
+
+module DBus
+  # better Type
+
+  # FIXME: in general, when an API gives me, a user, a choice,
+  # remember to make it easy for the case of:
+  # "I don't CARE, I don't WANT to care, WHY should I care?"
+  module Data
+    # Given a plain Ruby *value* and wanting a D-Bus *type*,
+    # construct an appropriate {Data::Base} instance.
+    #
+    # @param type [SingleCompleteType,Type]
+    # @param value [::Object]
+    # @return [Data::Base]
+    # @raise TypeError
+    def make_typed(type, value)
+      type = DBus.type(type) unless type.is_a?(Type)
+      data_class = Data::BY_TYPE_CODE[type.sigtype]
+      # not nil because DBus.type validates
+
+      data_class.from_typed(value, member_types: type.members)
+
+      # FIXME: TODO: implement
+    end
+    module_function :make_typed
+
+    # The base class for explicitly typed values.
+    #
+    # A value is either {Basic} or a {Container}.
+    # {Basic} values are either {Fixed}-size or {StringLike}.
+    class Base
```

Comment:
Just a general reflection, I tend to distrust deep hierarchies specially when they contain artificial intermediate classes like `StringLike`.

In this case, I can see it makes sense in the way it is. I just wanted to make a note that this activated my arachnid sense.

Reply:
Good point, the intermediate classes are either completely implementation detail (`Int`) or a direct transfer from the specification (`Fixed`, `StringLike`) and not really helpful for the user.

Would it help if their docs said "Abstract helper"?

Also, (module)`Data::`(class)`Base` could be eliminated and merged into `class Data`, similar to what I did in #104 with `Type::Type`.